File: app/src/main/java/com/loony/timelapsemaker/StorageManager.java
package com.loony.timelapsemaker;

import android.content.Context;
import android.os.Build;
import android.os.Environment;
import android.text.TextUtils;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Collections;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Created by Kamil on 7/23/2017.
 */

/*
Environment.getExternalStorageDirectory();                                      ->  /storage/emulated/0
Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);  ->  /storage/emulated/0/Pictures
*/

public class StorageManager {
    private static final String APPLICATION_STORAGE_PATH = "TimelapseMaker";
    private static final String REAL_EXT_SD_CARD_ABSOLUTE_PATH = "/storage/4542-1EE5/";

    private Context ctx;
    private File sdPath;

    public StorageManager(Context ctx) {
        this.ctx = ctx;

        File[] fs = ctx.getExternalFilesDirs(Environment.DIRECTORY_PICTURES);
        for(File f : fs) {
            if(!f.getAbsolutePath().contains("emulated"))
                sdPath = f;
        }
    }

    public boolean isExternalStorageAvailable() {
        String state = Environment.getExternalStorageState();
        return Environment.MEDIA_MOUNTED.equals(state);
    }

    public boolean isRealExternalStorageAvailable() {
//        boolean available = new File(REAL_EXT_SD_CARD_ABSOLUTE_PATH).exists();
//        return available;
        return sdPath != null;
    }

    /**
     *
     * @return  the directory name, or null if failed to create directory
     */
    public String createDirectory() {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss");
        String dirName = sdf.format(new Date());

        File absolutePath = new File(getPath(dirName));

        boolean canWrite = absolutePath.canWrite();
        boolean canRead = absolutePath.canRead();
        //Util.log("createDirectory -> " + dirName + " | " + absolutePath.getAbsolutePath());

        boolean creatingResult = absolutePath.mkdirs();

        if(creatingResult)
            return dirName;

        return null;
    }

    /**
     *
     * @param directory directory name
     * @param imageName image name without extension
     * @param imageBytes image
     * @return true if successfully saved, false if was any problem
     */
    public boolean saveImage(String directory, String imageName, byte[] imageBytes) {
        if(NewActivity.DEBUG_doNotSaveImageInStorageWhenCaptured)
            return true;

        File photo = new File(String.format("%s/%s.jpg", getPath(directory), imageName));

        try {
            FileOutputStream fos = new FileOutputStream(photo.getPath());
            fos.write(imageBytes);
            fos.close();
            Util.log("StorageManager::savingImage to " + photo.getAbsolutePath());
        } catch(IOException e) {
            Util.log("StorageManager::saveImage() exception -> " + e.getMessage());
            return false;
        }
        return true;
    }

    private String getPath(String directoryName) {
//        String normalPath = String.format("%s/%s/%s",
//                Environment.getExternalStorageDirectory(),
//                APPLICATION_STORAGE_PATH,
//                directoryName);

//        String customPathForExtSD = String.format("%s/%s/%s",
//                REAL_EXT_SD_CARD_ABSOLUTE_PATH,
//                APPLICATION_STORAGE_PATH,
//                directoryName);

        String customPathForExtSD = String.format("%s/%s",
                sdPath.getAbsolutePath(),
//                APPLICATION_STORAGE_PATH,
                directoryName);

        return customPathForExtSD;
    }

    private static final Pattern DIR_SEPARATOR = Pattern.compile("/");

}

File: app/src/main/java/pl/kflorczyk/timelapsemaker/StorageManager.kt
package pl.kflorczyk.timelapsemaker

import android.content.Context
import android.os.Environment
import android.util.Log
import pl.kflorczyk.timelapsemaker.Util.log
import java.io.File
import java.text.SimpleDateFormat
import java.util.*
import java.io.FileOutputStream
import java.io.IOException

class StorageManager(type: StorageType, context: Context) {
    private val TAG = "StorageManager"

    private val type: StorageType = type
    private val context: Context = context
    private val storageBase: File
    private var savingPath: File? = null

    init {
        val storages = getStorages(context)
        val storage = storages.find { p -> p.second == this.type } ?: throw RuntimeException("Storage not found for given type")
        storageBase = storage.first
    }

    fun createTimelapseDirectory(): Boolean {
        val sdf = SimpleDateFormat("yyyyMMdd_HHmmss")
        val dirName = sdf.format(Date())

        savingPath = File(getPath(dirName))
        return savingPath!!.mkdirs()
    }

    fun getPath(directoryName: String): String {
        return "${storageBase.absolutePath}/$directoryName/"
    }

    fun savePhoto(bytes: ByteArray, number: Int) {
        if(savingPath == null) {
            throw RuntimeException("Saving path is not specified (Did you invoke createTimelapseDirectory?)")
        }

        val photo = File(String.format("%s/photo%d.jpg", savingPath!!.absolutePath, number))

        try {
            val fos = FileOutputStream(photo.path)
            fos.write(bytes)
            fos.close()
            log(TAG, "savePhoto(ByteArray, Int), location = " + photo.absolutePath)
        } catch (e: IOException) {
            log(TAG, "savePhoto(ByteArray, Int) -> IOException:  " + e.message)
            throw e
        }
    }

    companion object {
        fun getStorages(context: Context): List<Pair<File, StorageType>> {
            val externalFilesDir = context.getExternalFilesDirs(Environment.DIRECTORY_PICTURES)

            val validate = fun(f: File): Boolean {
                val accessible = f.canWrite() && f.canRead()
                return accessible
            }

            val list = ArrayList<Pair<File, StorageType>>()

            for (f in externalFilesDir) {
                if (validate(f)) {
                    val type = if (f.absolutePath.contains("emulated")) StorageType.EXTERNAL_EMULATED else StorageType.REAL_SDCARD
                    list.add(Pair(f, type))
                }
            }

            return list
        }
    }

    enum class StorageType {
        EXTERNAL_EMULATED,
        REAL_SDCARD
    }
}